plan/logical_plan_builder.go

```diff
@@ -968,18 +976,284 @@ func (g *gbyResolver) Leave(inNode ast.Node) (ast.Node, bool) {
+					continue
+				}
+			}
+			if jCol, ok := joinDepends[iCol]; ok {
+				if _, ok = gbyCols[jCol]; ok {
+					continue
+				}
+			}
+			funcDepend = false
+			break
+		}
+		if funcDepend {
+			return true
+		}
+	}
+	primaryFuncDepend := true
+	hasPrimaryField := false
+	for _, colInfo := range tblInfo.Columns {
+		if !mysql.HasPriKeyFlag(colInfo.Flag) {
+			continue
+		}
+		hasPrimaryField = true
+		pCol, err := p.Schema().FindColumn(&ast.ColumnName{
+			Schema: col.DBName,
+			Table:  col.TblName,
+			Name:   colInfo.Name,
+		})
+		if err != nil {
+			primaryFuncDepend = false
+			break
+		}
+		if _, ok := gbyCols[pCol]; ok {
+			continue
+		}
+		if wCol, ok := whereDepends[pCol]; ok {
+			if _, ok = gbyCols[wCol]; ok {
+				continue
+			}
+		}
+		if jCol, ok := joinDepends[pCol]; ok {
+			if _, ok = gbyCols[jCol]; ok {
+				continue
+			}
+		}
+		primaryFuncDepend = false
+		break
+	}
+	return primaryFuncDepend && hasPrimaryField
+}
+
+func (b *planBuilder) checkOnlyFullGroupBy(p LogicalPlan, fields []*ast.SelectField, gby *ast.GroupByClause, from ast.ResultSetNode, where ast.ExprNode) {
+	gbyCols := make(map[*expression.Column]bool, len(fields))
+	gbyExprs := make([]ast.ExprNode, 0, len(fields))
+	schema := p.Schema()
+	for _, byItem := range gby.Items {
+		if colExpr, ok := byItem.Expr.(*ast.ColumnNameExpr); ok {
+			col, err := schema.FindColumn(colExpr.Name)
+			if err != nil {
+				continue
+			}
+			gbyCols[col] = true
+		} else {
+			gbyExprs = append(gbyExprs, byItem.Expr)
+		}
+	}
+
+	notInGbyCols := make(map[*expression.Column]int, len(fields))
+	for index, field := range fields {
+		if field.Auxiliary {
+			continue
+		}
+		if _, ok := field.Expr.(*ast.AggregateFuncExpr); ok {
+			continue
+		}
+		if _, ok := field.Expr.(*ast.ColumnNameExpr); !ok {
+			foundInGby := false
+			for _, expr := range gbyExprs {
+				if reflect.DeepEqual(expr, field.Expr) {
+					foundInGby = true
+					break
+				}
+			}
+			if foundInGby {
+				continue
+			}
+		}
+		colMap := make(map[*expression.Column]bool, len(schema.Columns))
+		b.allColFromExprNode(p, field, colMap)
+		for col := range colMap {
+			if _, ok := gbyCols[col]; ok {
+				continue
+			}
+			notInGbyCols[col] = index
+		}
+	}
+	if len(notInGbyCols) == 0 {
+		return
+	}
+
+	whereDepends := buildWhereFuncDepend(p, where)
+	joinDepends := buildJoinFuncDepend(p, from)
+	tblMap := make(map[*model.TableInfo]bool, len(notInGbyCols))
+	for col, index := range notInGbyCols {
+		tblInfo := tblInfoFromCol(from, col)
+		if tblInfo == nil {
+			continue
+		}
+		if _, ok := tblMap[tblInfo]; ok {
+			continue
+		}
+		if !checkColFuncDepend(p, col, tblInfo, gbyCols, whereDepends, joinDepends) {
+			b.err = ErrFieldNotInGroupBy.GenByArgs(index+1, fields[index].Text())
+			return
+		}
+		tblMap[tblInfo] = true
+	}
+
+}
+
+func (b *planBuilder) allColFromExprNode(p LogicalPlan, n ast.Node, cols map[*expression.Column]bool) {
+	switch v := n.(type) {
+	case *ast.BinaryOperationExpr:
```

Comment:
why not consider other exprs, like `ast.IsNullExpr` `ast.UnaryOperationExpr` ?

Reply:
sorry, I forgot to supplement other exprs.

Reply:
maybe we can implement a visitor to visit the ExprNode and store all columns ?

Reply:
yes, the implement of a visitor would be better